code/modules/atmospherics/machinery/pipes/multiz.dm

```diff
@@ -32,26 +32,31 @@
 /obj/machinery/atmospherics/pipe/multiz/SetInitDirections()
 	initialize_directions = dir
 
-/obj/machinery/atmospherics/pipe/multiz/update_layer()
-	return // Noop because we're moving this to /obj/machinery/atmospherics/pipe
-
 /obj/machinery/atmospherics/pipe/multiz/update_overlays()
 	. = ..()
+	cut_overlays()
 	pipe.color = front_node ? front_node.pipe_color : rgb(255, 255, 255)
 	pipe.icon_state = "pipe-[piping_layer]"
 	. += pipe
 	center.pixel_x = PIPING_LAYER_P_X * (piping_layer - PIPING_LAYER_DEFAULT)
 	. += center
-
+	update_layer()
 
 ///Attempts to locate a multiz pipe that's above us, if it finds one it merges us into its pipenet
 /obj/machinery/atmospherics/pipe/multiz/pipeline_expansion()
 	var/turf/T = get_turf(src)
-	var/obj/machinery/atmospherics/pipe/multiz/above = locate(/obj/machinery/atmospherics/pipe/multiz) in(SSmapping.get_turf_above(T))
-	var/obj/machinery/atmospherics/pipe/multiz/below = locate(/obj/machinery/atmospherics/pipe/multiz) in(SSmapping.get_turf_below(T))
-	if(below)
-		below.pipeline_expansion() //If we've got one below us, force it to add us on facebook
-	if(above)
-		nodes += above
-		above.nodes += src //Two way travel :)
+	var/list/above_adapters = list()
+	var/list/below_adapters = list()
+	for(var/obj/machinery/atmospherics/pipe/multiz/above in SSmapping.get_turf_above(T))
+		above_adapters += above
+	for(var/obj/machinery/atmospherics/pipe/multiz/below in SSmapping.get_turf_below(T))
+		below_adapters += below
+
+	for(var/obj/machinery/atmospherics/pipe/multiz/below in below_adapters)
+		if(below && below.piping_layer == piping_layer)
+			below.pipeline_expansion() //If we've got one below us, force it to add us on facebook
+	for(var/obj/machinery/atmospherics/pipe/multiz/above in above_adapters)
+		if(above && above.piping_layer == piping_layer)
+			nodes += above
+			above.nodes += src //Two way travel :)
```

Comment:
Haven't really sanity checked these, but at a very sleep deprived glance:
Why make a list of these if you're just gonna iterate over the same thing right after? Shouldn't this be done in the `for(multiz in SSmapping.get_turf())` loops rather then in a new one?

Reply:
yer right, i've changed it now